### ISSUE TITLE
Add callbacks for when user selects a voxel on a plot

### ISSUE THREAD

**Issue by mathias-sm** [nilearn/nilearn#2545]
When exploring a dataset it would be great to be able to make plots to which one can attach callbacks on voxel selection to perform arbitrary computation at that voxel's location.

My understanding is that the `view_img` route, by relying on `brainsprite.js`, is not the most suitable since the generated view will have a hard time communicating back events to the code (is that correct?).

Using `matplotlib`'s event handling features [I started to craft something](https://gist.github.com/mathias-sm/c8db344235c8a0922f5c60636c914f7c), but it looks like we can't use the `pick_event` callback and have to rely on `button_press_event` events — and I don't know how to convert from plot-coordinates (which depend on the window size, pixel density of the screen, and whatnot) to a more suitable system, because I don't understand how the images are blitted on the plot. My examples show almost what I want except for the `coords_from_event` function which I don't know how to implement.

Thanks,

**Comment by bthirion** [comment]
OK, I think that the way we've been handling this so far was to extract coordinates of the relevant voxels, and then create a dedicated masker object (e.g. spheres masker) to launch whatever analysis you like to launch.
I have no experience in capturing matplotlib events, converting them to positions, then real world coordinates etc.  and specifically I don't know how to test such features.
Any suggestion ? Should this go to a side project or do we really need this in nilearn ? 
My feeling is that interactivity should rather go into `view_*` functions. @pbellec, any hint ?
Best,

**Comment by pbellec** [comment]
In terms of `view_img` I agree with @mathias-sm it will not be possible to communicate events back to python. I am working on a new API for brainsprite in order to build dashboards that includes viewers, but it will be a mix of html templates (powered by brainsprite and plotly) and python. FYI the current API is described here brainsprite.github.io/ but about to undergo major renovations. I think it's a viable route to build interactive visualizations with complex operations, but it requires some familiarity with hmtl. FYI I am planning to submit PRs to nilearn in the future with that approach - starting with a 4D viewer with time series plotting. 

re whether a matplotlib viewer would belong in nilearn, I don't have a good sense of the long-term viability of a matplotlib approach, specifically whether it will be responsive, small (in memory and storage footprint) and extensible. The interactive visualization landscapes moves very fast in python, e.g. plotly's dash. So far nilearn has relied on old-school html python templating (not just in view_img, but also view_surface and the reports). I think it's a stable, proven solution regarding all the aspects I mentioned (responsiveness, footprint, extensions). In my perspective there would need a very strong proof-of-concept before seriously considering including a different approach.
 
@mathias-sm if you want to push this forward, the approach of view_img is to resample the data such that voxel coordinates can be extracted by simply using voxel size and origin. No cosines or anisotropic voxels: all that is taken care of by nilearn's resampling code prior to feeding in the vizualization. It should be possible to re-use many of the pieces of `view_img` to get this done, passing the info to matplotlib instead of brainsprite. It may even be possible to have a "render" option to select either brainsprite or a matplotlib one. Happy to point you to relevant bits of code if that's of interest.

**Comment by mathias-sm** [comment]
Update: after spending some time adapting ideas borrowed from `view_img`, I ended up (badly) reinventing the wheel. I had [something that worked](https://gist.github.com/mathias-sm/4838dc0456a031bbc5c904ebe1e37316#file-plot_img_with_callback3-py) for my use case but that was very ad hoc and didn't make use of the expertise, optimization and testing that went into the `plotting` module.

So [I started again](https://gist.github.com/mathias-sm/4838dc0456a031bbc5c904ebe1e37316#file-plot_img_with_callback_v0-0-1-py) with a class that warps around `plot_img`, intercepts the clicks and update the plots accordingly. I think it's a reasonable showcase of what the end-goal could look like, but it's still rough edged (e.g. will behave incorrectly on non-cubic voxels) and rather slow (it updates much more than needed with each click. Clicking and dragging is not really usable for example).

I'll start iterating on that, but I don't feel like my knowledge of either `pyplot` or the `plotting` module is good enough to suggest a reasonable first pull request. If someone with knowledge of the plotting code wants to chime in and give a few pointers at what looks reasonable in that first PoC and what needs to drastically change, and how to integrate this in the main codebase (including tests?) I'll be happy to push this forward!

**Comment by jeromedockes** [comment]
Congratulations on the new demo, I just tried it and it works very
smoothly!

I believe it may be more realistic to keep developing this as a separate
project for a while
(possibly in [nilearn sandbox](https://github.com/nilearn/nilearn_sandbox) )

At the moment is is not obvious what features would be needed by most
nilearn users or what interface would be effective. Also, approaches
involving a web browser such as plotly or jupyter widgets would need to
be compared against matplotlib -- I suspect the former may be easier and
more powerful tools for interactivity. Moreover, ATM even the static
plots are not tested as thoroughly as they should be IMHO, and testing
interactive user interfaces is harder.

Therefore I think adding this facet to nilearn would require significant
demand from users, and a relatively mature demo that has explored what
is useful, what works and what doesn't. It may be better adressed with a
standalone sister project (maybe one that can afford a dependency eg on
plotly, and that can evolve faster at first). Such a project could also
have better surface plots with the possibility to automatically save the
plotly figures as pngs, which ATM is an important limitation of
nilearn's 3D `view_*` functions.

I guess rather than thinking in terms of PRs and integration into
nilearn in the short term I would recommend continuing the way you have
started, ie as a separate module or library. Of course nilearn
developers with better knowledge of `nilearn.plotting` and matplotlib
are always happy to answer questions here or on the neurostars forum.

That is only my first impression, though, and I am curious re. what
other nilearn contributors think.

**Comment by bthirion** [comment]
Indeed, I was going to propose the same way forward: build an associated project. I know that this is not great for long-term maintenance, but good to make POCs.
Best,

**Comment by emdupre** [comment]
Would agree re initially developing separately. I just wanted to point to nibabel's `orthoview()` code, which could serve as a scaffolding. 

In particular, I know https://github.com/nipy/nibabel/pull/850 was started to add overlays to orthoviewer, and it sounds like could be useful as a comparison point !

**Comment by mathias-sm** [comment]
Thanks everyone for weighting in! I'm going keep my attempts separated, and specialized for my problems, for the moment.

With that said I kept exploring a bit, so I might as well [leave a trace here](https://gist.github.com/mathias-sm/0d9b4506c05baad1afdf0d6bb0130fe3). This is an attempt at doing it with `plotly`, and I think I'll iterate on that base for my work. It's not super snappy for high resolution data (for a single contrast `view_img` is far better) and I'm sure it's subtly wrong in various ways, but it should be fairly easy to adapt and reuse — and from now on it will probably just become more and more specialized for my use-case so it won't be worth sharing anymore.

**Comment by jeromedockes** [comment]
Thanks a lot for sharing, that's super interesting and indeed it is
useful to have a trace of it here.
On my laptop the plotly demo seems fast enough to use, I would say
around .25 seconds to update the plots when I click on a voxel.

I am not surprised that it is much slower than brainsprite, as it slices
and recomputes a heatmap for every update. (moreover the list
comprehension iterating over all values in a slice in `cut` is probably
quite costly). By contrast, brainsprite already prepares and embeds the
whole image in the page itself, so updates happen entirely in the
browser. If performance is a problem moving some of the work to the
browser may help; in this case as you are using dash you could have a
look at https://dash.plotly.com/clientside-callbacks

**Comment by bthirion** [comment]
Thx for sharing !

**Comment by mathias-sm** [comment]
Well, thanks again, it was pretty obvious but thresholding out of the loop makes quite a difference. I updated the gist.

I tried playing with `heatmapgl`, as well as with sending the background as a static image instead of a heatmap, but it doesn't seem speed things up. I can see how the next obvious improvement is going to be implementing the slices viewer on the client side (+ sending everything in one go), while letting the callbacks run on the server side. In fact, if I go that route I'll try to embed a `brainsprite` canvas pretty much directly.
